### PR TITLE
5333 dbsync test tool

### PR DESCRIPTION
**Pull request by tinachoB**
# Adding dbsync testing tool

## Issue 
#5333 

## Description
The main purpose of these changes are related to the addition of the dbsync testint tool. This one will be extremely useful to exercise the dbsync functionality in a deeper and strong way and, moreover, will be used in the continuous integration framework to automate all needed tests cases.

### CODE REVIEW

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    bool argsAreOK() const
+    {
+        bool res{ false };
+        if(m_cmdLineArgs.size() != 0)
+        {
+            const bool configOK
+            { 
+                m_cmdLineArgs[ConfigArg] == "-c" && !m_cmdLineArgs[ConfigValue].empty()
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:74]: if the arguments order change these checks will fail, is that ok?

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        ConfigArg       = 0,    // -c arg
+        ConfigValue     = 1,    // -c value
+        SnapshotsArg    = 2,    // -i arg
+        SnapshotsValue  = 3,    // -i value
+        OutputArg       = 4,    // -o arg
+        OutputValue     = 5     // -o value
+    };
+
+    CmdLineArgs() = delete;
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:99]: not needed.

File: src/dbsync/testtool/main.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
+/*
+ * Wazuh DBSYNC
+ * Copyright (C) 2015-2020, Wazuh Inc.
+ * June 11, 2020.
```
Comment (konangc87) [inline src/dbsync/testtool/main.cpp:4]: wrong date

File: src/dbsync/testtool/main.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        {
+            // Let's take the input json list and apply the changes to the db
+            for (size_t idx = 0; idx < snapshots.size(); ++idx) 
+            {
+                const std::string snapshotsIdxFile{ snapshots[idx] };
+                const std::unique_ptr<cJSON, CJsonDeleter> currentSnapshot
+                { 
+                    cJSON_Parse(currentSnapToString(snapshotsIdxFile).c_str())
+                };
+                cJSON* snapshotLambda{ nullptr };
```
Comment (konangc87) [inline src/dbsync/testtool/main.cpp:82]: use a unique_ptr with a deleter calling dbsync_free_result.

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    CmdLineArgs(const int argc, char* argv[])
+    {
+        m_cmdLineArgs.reserve(argc);
+        for (int i = 1; i < argc; ++i)
+        {
+            m_cmdLineArgs.push_back(argv[i]);
+        }        
+    }
+
+    std::string configFile() const
```
[the remark this answers is not in the document]
Comment (tinachoB) [reply src/dbsync/testtool/cmdArgsHelper.h:32]: good catch

File: src/dbsync/include/typedef.h [hunk 6]
```diff
@@ -21,16 +21,20 @@
...
+{
     MANAGER = 0,
     AGENT = 1
 }HostType;
 
-typedef enum {
-    SQLITE3 = 0
+typedef enum 
+{
+    UNDEFINED = -1,
```
[the remark this answers is not in the document]
Comment (tinachoB) [reply src/dbsync/include/typedef.h:32]: We'll update the Coding Style with this ;)

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        bool res{ false };
+        if(m_cmdLineArgs.size() != 0)
+        {
+            bool configOK{ false };
+            bool snapshotsOK{ false };
+            bool outputOK{ false };
+            for(unsigned int i = 0; i < m_cmdLineArgs.size(); ++i)
+            {
+                if(m_cmdLineArgs[i].compare("-c") == 0 && !m_cmdLineArgs[i+1].empty())
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:77]: if there is no value after i, the i+1 will end in an overflow. You should check if i+1 < m_cmdLineArgs.size() before using the index.

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                {
+                    configOK = true;
+                }
+                else if(m_cmdLineArgs[i].compare("-s") == 0 && !m_cmdLineArgs[i+1].empty())
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:81]: ditto

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+                {
+                    snapshotsOK = true;
+                }
+                else if(m_cmdLineArgs[i].compare("-o") == 0 && !m_cmdLineArgs[i+1].empty())
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:85]: ditto

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+                {
+                    outputOK = true;
+                }
+            }
+            res = configOK && snapshotsOK && outputOK;
+        }
+        return res;
+    }
+private:
+    enum CmdLineValues
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:95]: I think this enum is not needed anymore.

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        { 
+            std::string substr; 
+            getline(ss, substr, ','); // Getting each string between ',' character 
+            snapshots.push_back(std::move(substr));
+        } 
+    }
+
+    const std::string& outputFolder() const
+    {
+        return m_cmdLineArgs[OutputValue];
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:51]: you are assuming that the output folder value will be always in the OutputValue index and that's always true.

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        m_cmdLineArgs.reserve(argc);
+        for (int i = 1; i < argc; ++i)
+        {
+            m_cmdLineArgs.push_back(argv[i]);
+        }        
+    }
+
+    const std::string& configFile() const
+    {
+        return m_cmdLineArgs[ConfigValue];
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:34]: the same here.

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+#include <vector>
+#include <iostream>
+
+class CmdLineArgs
+{
+public:
+    CmdLineArgs(const int argc, const char* argv[])
+    {
+        m_cmdLineArgs.reserve(argc);
+        for (int i = 1; i < argc; ++i)
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:26]: in here you can check if the number of parameters is the expected one and parse them.

File: src/dbsync/testtool/main.cpp [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+                if(0 == dbsync_update_with_snapshot(handle, currentSnapshot.get(), &snapshotLambda))
+                {
+                    // Create and flush snapshot diff data in files like: snapshot_<#idx>.json
+                    std::cout << "Processing file: " << snapshots[idx] << std::endl;
+                    std::stringstream oFileName;
+                    oFileName << "snapshot_" << idx << ".json";
+                    const std::string outputFileName{ cmdLineArgs.outputFolder()+"/"+oFileName.str() };
+                    std::ofstream outputFile{ outputFileName };
+                    const std::unique_ptr<char, CharDeleter> snapshotDiff{ cJSON_Print(snapshotLambda) };
```
Comment (konangc87) [inline src/dbsync/testtool/main.cpp]: you should use dbsync_free_result as deleter function.

File: src/dbsync/testtool/main.cpp [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+                    outputFile << snapshotDiff.get() << std::endl;
+                    outputFile.close();
+                    dbsync_free_result(&snapshotLambda);
```
Comment (konangc87) [inline src/dbsync/testtool/main.cpp]: double free?

Comment (tinachoB) [reply src/dbsync/testtool/main.cpp]: I think you are seeing an outdated version :)

File: src/dbsync/testtool/Readme.md [hunk 16]
```diff
@@ -0,0 +1,43 @@
+# DBSync Testing Tool
+## Index
+1. [Purpose](#purpose)
+2. [Compile Wazuh](#compile-wazuh)
+3. [How to use the tool](#how-to-use-the-tool)
+
+## Purpose
+The DBSync Testing Tool was created to test and validate the correct
```
Comment (konangc87) [inline src/dbsync/testtool/Readme.md:8]: I guess there is something missing here.

File: src/dbsync/testtool/Readme.md [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    "db_name": "db_name",
+    "db_type": "1",
+    "host_type": "<0|1>",
+    "persistance": "",
+    "sql_statement":"sql"
+}
+```
+Where:
+  - db_name: Database name to be used.
+  - db_type: Database type to be used. Right now it is being used SQLITE3.
```
Comment (konangc87) [inline src/dbsync/testtool/Readme.md:31]: "Only SQLITE3 is currently supported"

File: src/dbsync/testtool/Readme.md [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+  - host_type: Agent or Manager.
+  - persistance: Database type of persistance being used. Not implemented yet.
+  - sql_statement: Database sql structure to be created. This structure will be associated the other files needed to used the tool.
+
+2) Create the needed amount of json files representing the different snapshots information. These ones need to follow the sql_statement structure created in the below step.
```
Comment (konangc87) [inline src/dbsync/testtool/Readme.md:36]: in the step above or in step # 1

File: src/dbsync/testtool/Readme.md [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+3) Define an output folder where all resulting data will be located.
+4) Once all the below steps are accomplished the tool will be used like this:
```
Comment (konangc87) [inline src/dbsync/testtool/Readme.md:38]: above

File: src/dbsync/testtool/Readme.md [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+```
+./dbsync_test_tool -c config.json -s input1.json,input2.json,input3.json -o ./output
+```
+5) Considering the example below all diff snapshots will be located in ./output folder in the following format: snapshop_1.json, snapshot_2.json ... snapshot_n.json where 'n' will be the number of json files passed as part of the argument "-s".
```
Comment (konangc87) [inline src/dbsync/testtool/Readme.md:42]: above

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+        const auto argsSize{ m_cmdLineArgs.size() };
+        for(unsigned int i = 0; i < argsSize; ++i)
+        {
+            if(m_cmdLineArgs[i].compare(input) == 0 && (i+1 < argsSize) && !m_cmdLineArgs[i+1].empty())
+            {
+                result = m_cmdLineArgs[i+1];
+                break;
+            }
+        }
+        return std::move(result);
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:108]: remove the const from the return and the std::move won't be needed.

File: src/dbsync/testtool/cmdArgsHelper.h [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    std::vector<std::string> m_cmdLineArgs;
```
Comment (konangc87) [inline src/dbsync/testtool/cmdArgsHelper.h:111]: I think you can have something like:
const std::vector<std::string> m_snapshots;
const std::string m_configFileName;
const std::string m_outputFolder;
and fill their values in the ctor using private functions that look for the switch in each case and throw an exception if the switch or the value is not present.

File: src/dbsync/testtool/main.cpp [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+                }
+            }
+            std::cout << "Resulting files are located in the "<< cmdLineArgs.outputFolder() << "folder" << std::endl;
+        }
+        else
+        {
+            std::cout << "\nSomething went wrong configuring the database. Please, check the config file data" << std::endl;
+            CmdLineArgs::showHelp();
+        }
+    }
```
Comment (Dwordcito) [inline src/dbsync/testtool/main.cpp:107]: could you call dbsync_teardown here?

Comment (tinachoB) [reply src/dbsync/testtool/main.cpp:107]: Yes sir!!

**Pull request merged**